File: PHP/generateZip.php
<?php
// Log Document
function _log($str)
{
    // log to the output
    $log_str = date('d.m.Y') . ": {$str}\r\n";
    echo $log_str;

    // log to file
    if (($fp = fopen('upload_log.txt', 'a+')) !== false) {
        fputs($fp, $log_str);
        fclose($fp);
    }
}

# This PHP code grabs the specified files by the user, creates and copys them to a temporary directory which is then ziped.

#Value of items consist of a large string with multiple variables delimited my "::". See Get List functions in main JavaScript file
     #$_GET["project"] = <Project name>::
     $selected_project = str_replace(".", "", $_GET["project"]);      #string manipulation to take our periods and replace spaces with underscores (_).
     $selected_project = str_replace(" ", "_", $selected_project);

     $selected_project_parts = explode("::", $selected_project);           #string manipulation to grab the project name
     $selected_project_name = $selected_project_parts[0];

     #$_GET['orthomosaic'] = <File nmme>::<path to file>::<EPSG value>
     $selected_orthomosaic_list = $_GET["orthomosaic"]; #string that contains orthomosiacs name, Path to file and orthomosaic's EPSG value.
     //$selected_orthomosaic_list = explode(",", $selected_orthomosaic);
     $ortho_list_length = count($selected_orthomosaic_list);
     $returning2dArray = array();
     for ($x = 0; $x < $ortho_list_length; $x++) {
        $selected_orthomosaic = $selected_orthomosaic_list[$x];
        $selected_orthomosaic_parts = explode("::", $selected_orthomosaic);                 #string manipulation to grab orthomosaics name w/o extension.
        $selected_orthomosaic_FileName = explode(".", $selected_orthomosaic_parts[0]);
        $selected_orthomosaic_FileName_noExt = $selected_orthomosaic_FileName[0];

        $selected_CanopyHeightModel_list = $_GET["chm"];
        //$selected_CanopyHeightModel_list = explode(",", $selected_CanopyHeightModel);
        $chm_list_length = count($selected_CanopyHeightModel_list);
        for ($y = 0; $y < $chm_list_length; $y++) {
           //$selected_CanopyHeightModel = $_GET["chm"] ?? "0::0"; //0 if not selected to allow $generateZip_command from executing.
           //$selected_CanopyHeightModel = "0::0"; //0 if not selected to allow $generateZip_command from executing.
           $selected_CanopyHeightModel = $selected_CanopyHeightModel_list[$y];
           $selected_boundary = $_GET["shape"];

           // _log('Show selected project: '.$selected_project);
           // _log('Show selected orthomosaic: '.$selected_orthomosaic);
           // _log('Show selected CHM: '.$selected_CanopyHeightModel);
           // _log('Show selected boundary: '.$selected_boundary);

           $cc_checked = $_GET["cc_ckd"];       // check what atribute checkboxes are checked
           $exg_checked = $_GET["exg_ckd"];

           if($selected_CanopyHeightModel == "0"){         // if a CHM is not specified, set the state of the Canopy Height and Canopy Volume checkboxes to False
             $ch_checked = "false";
             $cv_checked = "false";
           }else{
             $ch_checked = $_GET["ch_ckd"];                // if a CHM is specified, grab the Canopy Height and Canopy Volume checkboxes states.
        //     array_push($returningArray, $ch_checked);
             $cv_checked = $_GET["cv_ckd"];
        //     array_push($returningArray, $cv_checked);
           }

           $csv_checked = $_GET["csv_ckd"];             // check what file format checkboxes are checked
           $xls_checked = $_GET["xls_ckd"];
           $geojson_checked = $_GET["geo_ckd"];
           $shape_checked = $_GET["shp_ckd"];

           $returningArray = array();
           array_push($returningArray, $selected_project);
           array_push($returningArray, $selected_orthomosaic);
           array_push($returningArray, $selected_boundary);

           # Call Python code with field information to generate a Zip file with the results.
           $generateZip_command = "python3 /var/www/html/uas_tools/canopy_attribute_generator/Resources/Python/generateZip.py $selected_project $selected_orthomosaic $selected_CanopyHeightModel $selected_boundary $csv_checked $xls_checked $geojson_checked $shape_checked $cc_checked $exg_checked $ch_checked $cv_checked";
           array_push($returningArray, $generateZip_command);
           $result = shell_exec($generateZip_command);
           _log('Command to generate zip file : '.$generateZip_command);
           //_log('Result from command : '.$result);



           if($result){                                     // If there is a result push result to array being passed back in JSON format
             # echo "Python has been executed!";
             array_push($returningArray, $result);
              }else{
                array_push($returningArray, "ERROR: Was unable to call to the generateZip python executable!");     // Else, push Error on to returning JSON object.
              }
              array_push($returning2dArray, $returningArray);
            }
        }

     _log('Generate Zip:');
     _log('Show selected project: '.$selected_project);
     _log('Show selected orthomosaic: '.$selected_orthomosaic);
     _log('Show selected CHM: '.$selected_CanopyHeightModel);
     _log('Show selected boundary: '.$selected_boundary);
     echo json_encode($returning2dArray);     //return returning array regarding this call in JSON format.
     die();

?>

File: Python/generateZip.py
import sys
import os
from zipfile import ZipFile
import shutil
import time

def main():
    #string manipulation of the selected values of the dropdown menus to grab variables needed to calculate attributes.
    # the parameters sent in are long strings containing meta data of each selected field (project, orthomosaic, boundary, etc.) each attribute of selected field split with "::".

    selected_project = sys.argv[1].split("::")
    selected_project_name = selected_project[0]                                              # name of the selected project

    selected_orthomosaic = sys.argv[2].split("::")
    selected_orthomosaic_FileName = selected_orthomosaic[0]                                      # File name of the selected orthomosaic with file extension
    selected_orthomosaic_FilePath = selected_orthomosaic[1]                                     # file path of the selected orthomosaic
    selected_orthomosaic_EPSG = selected_orthomosaic[2]                                          # EPSG value of the selected orthomosaic
    selected_orthomosaic_FileName_noExt = selected_orthomosaic_FileName.split(".")
    selected_orthomosaic_FileName_noExt = selected_orthomosaic_FileName_noExt[0]                #  file name of selected orthomosaic withoout file extension
    print("!!!!!!!!!!!!!!!!", selected_orthomosaic_FileName)

    if(sys.argv[3] == "0"):
        selected_CanopyHeightModel = sys.argv[3]
        print("There was no CHM selected!")
    else:
        selected_CanopyHeightModel = sys.argv[3].split("::")
        selected_CanopyHeightModel_FileName = selected_CanopyHeightModel[0]                              # File name of the selected Canopy Height Model (CHM) with file extension
        selected_CanopyHeightModel_FilePath = selected_CanopyHeightModel[1]                                 # file path of the selected CHM
        selected_CanopyHeightModel_FileName_noExt = selected_CanopyHeightModel_FileName.split(".")
        selected_CanopyHeightModel_FileName_noExt = selected_CanopyHeightModel_FileName_noExt[0]        # file name of selected CHM withoout file extension


    selected_boundary = sys.argv[4].split("::")
    selected_boundary_FileName = selected_boundary[0]                                           # File name of the selected boundary file with file extension
    selected_boundary_FilePath = selected_boundary[1]                                            # file path of the selected boundary file
    selected_boundary_FileName_noExt = selected_boundary_FileName.split(".")
    selected_boundary_FileName_noExt = selected_boundary_FileName_noExt[0]                          # file name of selected boundary file withoout file extension

    if(sys.argv[5] == "true"):              # grab checkbox parameters and set them to boolean.
        csv_checked = True
    else:
        csv_checked = False

    if(sys.argv[6] == "true"):
        xls_checked = True
    else:
        xls_checked = False

    if(sys.argv[7] == "true"):
        geojson_checked = True
    else:
        geojson_checked = False

    if(sys.argv[8] == "true"):
        shape_checked = True
    else:
        shape_checked = False

    if(sys.argv[9] == "true"):
        cc_checked = True
    else:
        cc_checked = False

    if(sys.argv[10] == "true"):
        exg_checked = True
    else:
        exg_checked = False

    if(sys.argv[11] == "true"):
        ch_checked = True
    else:
        ch_checked = False

    if(sys.argv[12] == "true"):
        cv_checked = True
    else:
        cv_checked = False

#base path to where attribute generation results are stored in regards to what project, Orthomosaic and boundary are selected.
    path_to_project_folder = "/var/www/html/uas_data/download/product/" + selected_project_name + "/" + selected_orthomosaic_FileName_noExt + "_" + selected_boundary_FileName_noExt + "/"
    if not (os.path.exists(path_to_project_folder)):
        print("The specified project/orthomosaic results do not exist!")
        return 0

#attributes array containing attribute boundary string extentions
#    attributes = ["cc_boundary", "exg_boundary", "ch_boundary", "cv_boundary"]

    attributes = []
    if(cc_checked):
        attributes.append("cc_boundary")
    if(exg_checked):
        attributes.append("exg_boundary")
    if(ch_checked):
        attributes.append("ch_boundary")
    if(cv_checked):
        attributes.append("cv_boundary")


#path to temporary directory that will hold the results
    #path_to_temp_folder = path_to_project_folder + selected_project_name + "_" + selected_orthomosaic_FileName_noExt + "_" + selected_boundary_FileName_noExt + "_results"
    path_to_temp_folder = path_to_project_folder
    temp_folder_exists = os.path.isdir(path_to_temp_folder)
    print("temp_folder_exists: " + str(temp_folder_exists))
'''
    if not (temp_folder_exists):                        #if the temp direcory doesn't exist create one
        os.makedirs(path_to_temp_folder)
        print("temp folder did not exist yet, creating one.")
    else:
        shutil.rmtree(path_to_temp_folder)              # if the temp directory exists, delete it then create the temporary directory.
        os.makedirs(path_to_temp_folder)
        print("temp folder already existed. removing and creating a new one.")
'''        
    index = 0
    while index < len(attributes):              # index through the number of attribute boundary string extensions
        if(csv_checked):
            file_path = path_to_project_folder + attributes[index] + "/" + attributes[index] + "_" + selected_boundary_FileName_noExt + ".csv"          #if the csv checkbox was checked. Copy the file to the temporary results directory.
            path_to_temp_folder += attributes[index]# + "/" + attributes[index] + "_" + selected_boundary_FileName_noExt + ".csv"
            temp_folder_exists = os.path.isdir(path_to_temp_folder)
            print("temp_folder_exists2: " + str(temp_folder_exists))
            print("@@@@@@@@@@@@", file_path)
            if os.path.exists(file_path):
                shutil.copy(file_path, path_to_temp_folder)
            else:
                print(file_path)
                print("The specified canopy attribute results do not exist!")                  #if not, state that the file does not exist.
                shutil.rmtree(path_to_temp_folder)
                return 0

        if(xls_checked):
            file_path = path_to_project_folder + attributes[index] + "/" + attributes[index] + "_" + selected_boundary_FileName_noExt + ".xlsx"           #if the xls checkbox was checked. Copy the file to the temporary results directory.
            if os.path.exists(file_path):
                shutil.copy(file_path, path_to_temp_folder)
            else:
                print(file_path)
                print("The specified canopy attribute results do not exist!")                  #if not, state that the file does not exist.
                shutil.rmtree(path_to_temp_folder)
                return 0

        if(geojson_checked):
            file_path = path_to_project_folder + attributes[index] + "/" + attributes[index] + "_" + selected_boundary_FileName_noExt + ".geojson"       #if the geojson checkbox was checked. Copy the file to the temporary results directory.
            if os.path.exists(file_path):
                shutil.copy(file_path, path_to_temp_folder)
            else:
                print(file_path)
                print("The specified canopy attribute results do not exist!")                  #if not, state that the file does not exist.
                shutil.rmtree(path_to_temp_folder)
                return 0

        if(shape_checked):
            file_path = path_to_project_folder + attributes[index] + "/" + attributes[index] + "_" + selected_boundary_FileName_noExt + ".shp"           #if the shape checkbox was checked. Copy the file to the temporary results directory.
            if os.path.exists(file_path):
                shutil.copy(file_path, path_to_temp_folder)
            else:
                print(file_path)
                print("The specified canopy attribute results do not exist!")                  #if not, state that the file does not exist.
                shutil.rmtree(path_to_temp_folder)
                return 0

            file_path = path_to_project_folder + attributes[index] + "/" + attributes[index] + "_" + selected_boundary_FileName_noExt + ".shx"           #if the shape checkbox was checked. Copy the file to the temporary results directory.
            if os.path.exists(file_path):
                shutil.copy(file_path, path_to_temp_folder)
            else:
                print("File does not exist: " + file_path)                  #if not, state that the file does not exist.

            file_path = path_to_project_folder + attributes[index] + "/" + attributes[index] + "_" + selected_boundary_FileName_noExt + ".dbf"           #if the shape checkbox was checked. Copy the file to the temporary results directory.
            if os.path.exists(file_path):
                shutil.copy(file_path, path_to_temp_folder)
            else:
                print("File does not exist: " + file_path)                  #if not, state that the file does not exist.
        index += 1
        print("index: " + str(index))

    path_to_zip = path_to_temp_folder + ".zip"                                      # set name of Zip file
    if(os.path.exists(path_to_zip)):                                                # if the Zip file already exists, delete it before creating it.
        os.remove(path_to_zip)
        shutil.make_archive(path_to_temp_folder, 'zip', path_to_temp_folder)
    else:
        shutil.make_archive(path_to_temp_folder, 'zip', path_to_temp_folder)



#---------------------------------------------------------------------------------------------------------------------Run Main
if __name__ == "__main__":
    main()
